### PR TITLE
Expose API for merkle hash

### PR DESCRIPTION
**Pull request by ebfull**

### CODE REVIEW

File: src/rustzcash.rs [hunk 1]
```diff
@@ -1,7 +1,100 @@
...
+    let tmp = pedersen_hash::<Bls12, _>(
+        Personalization::MerkleTree(depth),
+        lhs.iter().map(|&x| x)
+            .take(Fr::NUM_BITS as usize)
+            .chain(rhs.iter().map(|&x| x).take(Fr::NUM_BITS as usize)),
+        &params.jubjub_params
+    ).into_xy().0.into_repr();
+
+    // Should be okay, caller is responsible for ensuring the pointer
+    // is valid.
```
Comment (str4d) [inline src/rustzcash.rs]: `is a valid pointer to 32 bytes.`

File: src/rustzcash.rs [hunk 2]
```diff
@@ -1,7 +1,100 @@
...
+    for (a, b) in lhs.iter_mut().rev().zip(BitIterator::new(a_repr)) {
+        *a = b;
+    }
+
+    for (a, b) in rhs.iter_mut().rev().zip(BitIterator::new(b_repr)) {
+        *a = b;
+    }
+    
+    let tmp = pedersen_hash::<Bls12, _>(
+        Personalization::MerkleTree(depth),
```
Comment (str4d) [inline src/rustzcash.rs]: What happens if a depth larger than 62 is passed in? Relatedly, how do panics unroll into the C++ code - as segfaults?

Comment (ebfull) [reply src/rustzcash.rs]: If a depth larger than 62 is passed in, undefined behavior. Probably a segfault.

**Comment by ebfull** [comment]
I configured it to abort the process on panics, rather than causing chaos downstream over the FFI boundary. Also enabled link-time optimizations so the static lib is smaller.

File: src/rustzcash.rs (hunk 2 above)
Comment (str4d) [reply src/rustzcash.rs]: Could we add a `panic!()` if the provided depth is larger than 62?

File: src/rustzcash.rs [hunk 3]
```diff
@@ -1,7 +1,97 @@
...
+        pedersen_hash,
+        Personalization
+    }
+};
+
+use libc::{uint64_t, size_t, c_uchar};
+
+lazy_static! {
+    static ref JUBJUB: JubjubBls12 = {
+        JubjubBls12::new()
```
Comment (str4d) [inline src/rustzcash.rs]: Just to confirm, this is only the Jubjub parameters? The Sapling and Sprout CRSs will be loaded/unloaded via pointers? Or will the Sapling CRS be similarly statically-loaded?

File: src/rustzcash.rs (hunk 2 above)
Comment (ebfull) [reply src/rustzcash.rs]: The code this calls will panic if the depth is larger then 62.

File: src/rustzcash.rs (hunk 3 above)
Comment (str4d) [reply src/rustzcash.rs]: @ebfull discussed this. The Sapling and Sprout CRSs will be loaded via an API through which zcashd passes the paths to the params. Both verification keys along with the Sapling proving key will be loaded into memory immediately, while the Sprout proving key will be loaded and unloaded on-demand.

**Comment by str4d** [comment]
@daira ACKed https://github.com/zcash/zcash/pull/3169 which includes this.

**Pull request merged**